turbopuffer/backend.py

```diff
@@ -140,14 +140,18 @@ def make_api_request(self,
                         raise_api_error(response.status_code, content.get('status', 'error'), content.get('error', ''))
                 else:
                     raise_api_error(response.status_code, 'Server returned non-JSON response', response.text)
-            except requests.HTTPError as http_err:
+            except (requests.HTTPError, requests.ConnectionError, requests.Timeout) as err:
                 retry_attempt += 1
                 # print(traceback.format_exc())
                 if retry_attempt < tpuf.max_retries:
                     # print(f'Retrying request in {2 ** retry_attempt}s...')
                     time.sleep(2 ** retry_attempt)  # exponential falloff up to 64 seconds for 6 retries.
                 else:
                     print(f'Request failed after {retry_attempt} attempts...')
-                    raise_api_error(http_err.response.status_code,
-                                   f'Request to {http_err.request.url} failed after {retry_attempt} attempts',
-                                   str(http_err))
+
+                    if isinstance(err, requests.HTTPError):
+                        raise_api_error(err.response.status_code,
+                                   f'Request to {err.request.url} failed after {retry_attempt} attempts',
+                                   str(err))
+                    else:
+                        raise
```

Comment:
is there any difference between `raise` and `raise err`?

Reply:
Not 100% sure but I believe it's the same as in Ruby, i.e., `raise` lets the exception that was caught bubble up again while `raise e` raises a "new" exception and overwrites the backtrace that `e` had before.